#!/bin/bash
#PBS -l select=1:ncpus=48:mpiprocs=1:ompthreads=48
#PBS -l walltime=24:00:00

# Initialise conda
source ~/.bashrc
conda activate deeptools-env

## Load conda environment with agat
conda activate agat-env

# Create GeneSets directory and navigate to it
mkdir -p $HOME/MultiomeProject/deeptools_Analysis/GeneSets
cd $HOME/MultiomeProject/deeptools_Analysis/GeneSets/

# Copy GeneSets OC1UGs.txt and WSC1UGs.txt created in R to this directory (if not already done)

# Filter the GFF for gene sets
agat_sp_filter_feature_from_keep_list.pl --gff $HOME/Physcogenome/Version6/P.patens_genome_annotation_v6.gff --keep_list OC1UGs.txt --output OC1UGs_filtered.gff
agat_sp_filter_feature_from_keep_list.pl --gff $HOME/Physcogenome/Version6/P.patens_genome_annotation_v6.gff --keep_list WSC1UGs.txt --output WSC1UGs_filtered.gff

# Convert filtered GFF files to GTF format
agat_convert_sp_gff2gtf.pl --gff OC1UGs_filtered.gff -o OC1UGs_filtered.gtf
agat_convert_sp_gff2gtf.pl --gff WSC1UGs_filtered.gff -o WSC1UGs_filtered.gtf

# Switch back to deeptools environment
conda deactivate
conda activate deeptools-env

# Create output directories for matrices and figures
mkdir -p $HOME/MultiomeProject/deeptools_Analysis/Matrices
mkdir -p $HOME/MultiomeProject/deeptools_Analysis/Figures

ls $HOME/MultiomeProject/deeptools_Analysis/BigWigs/

# Process each pair of BigWig files based on their naming pattern
for i in Cluster_1 Cluster_7 Cluster_Other; do
  wt_bw_file="$HOME/MultiomeProject/deeptools_Analysis/BigWigs/${i}_Wild_type-TileSize-100-normMethod-rc.bw"
  ste_bw_file="$HOME/MultiomeProject/deeptools_Analysis/BigWigs/${i}_Δstemin-TileSize-100-normMethod-rc.bw"

  echo "Processing Wild Type: $wt_bw_file and ∆stemin: $ste_bw_file"
  
  computeMatrix scale-regions -S "$wt_bw_file" "$ste_bw_file" \
                              -R $HOME/MultiomeProject/deeptools_Analysis/GeneSets/OC1UGs_filtered.gtf $HOME/MultiomeProject/deeptools_Analysis/GeneSets/WSC1UGs_filtered.gtf \
                              --beforeRegionStartLength 3000 \
                              --regionBodyLength 5000 \
                              --afterRegionStartLength 3000 \
                              --skipZeros -o $HOME/MultiomeProject/deeptools_Analysis/Matrices/${i}_Wild_type_vs_Δstemin_matrix.gz

  # Path to the matrix file generated by computeMatrix
  matrix_file="$HOME/MultiomeProject/deeptools_Analysis/Matrices/${i}_Wild_type_vs_Δstemin_matrix.gz"
  
  # Path to save the plot
  output_file="$HOME/MultiomeProject/deeptools_Analysis/Figures/${i}_profile.svg"

  # Run plotProfile
  plotProfile -m "$matrix_file" -out "$output_file" --yMin 0 --yMax 0.05

  echo "Profile plot generated for: $matrix_file"
done
